<a href="https://colab.research.google.com/github/cmshebeeb/SMFS-AutoEncoder/blob/main/notebooks/06_structure_preservation_laplacian_colon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Setup

In [1]:
from google.colab import userdata
github_token=userdata.get('GITHUB_TOKEN')
print("GitHub token loaded...")

GitHub token loaded...


### Clone GitHub Repository

In [2]:
import os

username = "cmshebeeb"
repo = "SMFS-AutoEncoder"

!git clone https://{username}:{github_token}@github.com/{username}/{repo}.git
print("Cloning Succesfull..")

Cloning into 'SMFS-AutoEncoder'...
remote: Enumerating objects: 152, done.
remote: Counting objects: 100% (152/152), done.
remote: Compressing objects: 100% (99/99), done.
remote: Total 152 (delta 52), reused 131 (delta 35), pack-reused 0 (from 0)
Receiving objects: 100% (152/152), 8.85 MiB | 11.72 MiB/s, done.
Resolving deltas: 100% (52/52), done.
Cloning Succesfull..


### Change to Project Folder

In [3]:
%cd SMFS-AutoEncoder

/content/SMFS-AutoEncoder


In [4]:
%pwd

'/content/SMFS-AutoEncoder'

### Install Dependencies

In [ ]:
!pip install -r requirements.txt

## Git Setup

### Configure Git

In [36]:
!git config --global user.name "cmshebeeb"
!git config --global user.email "mshebeebc@gmail.com"

# WORK SPACE

In [27]:
!git pull

remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 12 (delta 8), reused 12 (delta 8), pack-reused 0 (from 0)
Unpacking objects: 100% (12/12), 101.37 KiB | 506.00 KiB/s, done.
From https://github.com/cmshebeeb/SMFS-AutoEncoder
   4798c28..1dd3af8  main       -> origin/main
Updating 4798c28..1dd3af8
Fast-forward
 notebooks/01_eda_colon.ipynb                   |  33 ++-
 notebooks/05_l21_feature_selection_colon.ipynb | 381 ++++++++++++++++---------
 reports/figures/colon_acc_vs_k.png             | Bin 15128 -> 15130 bytes
 reports/figures/colon_nmi_vs_k.png             | Bin 15128 -> 15130 bytes
 reports/results/accuracy_results.csv           |   8 +-
 reports/results/feature_selection_results.csv  |   8 +-
 reports/results/nmi_results.csv                |   8 +-
 7 files changed, 284 insertions(+), 154 deletions(-)


In [5]:
!ls src/graph

laplacian.py


In [39]:
!ls notebooks

01_eda_colon.ipynb	 04_baseline_autoencoder.ipynb
02_eda_binary.ipynb	 05_l21_feature_selection_colon.ipynb
03_eda_multiclass.ipynb


In [6]:
from src.graph.laplacian import compute_laplacian

In [7]:
print(compute_laplacian)

<function compute_laplacian at 0x7842946177e0>


Function is correctly imported and available in Colab

### Data

In [9]:
import pandas as pd
import torch
from torch.utils.data import DataLoader, TensorDataset

In [10]:
df = pd.read_csv("data/preprocessed/colon.csv")

X = df.drop(columns=["label"])
X = torch.tensor(X.values, dtype=torch.float32)

In [11]:
dataset = TensorDataset(X)
train_loader = DataLoader(
    dataset,
    batch_size=16,
    shuffle=True
)

In [12]:
print("X shape:", X.shape)

X shape: torch.Size([62, 2000])


### Add the Laplacian Structure Loss

here need to compute the Lpaclacian of original input data and latent representation

In [13]:
# Input-data Laplacian

X_np = X.cpu().numpy()

L_input = compute_laplacian(X_np)

print("Input Laplacian shape:", L_input.shape)

Input Laplacian shape: (62, 62)


In [14]:
# Structure-preservation hyperparameter

lambda_graph = 0.001

print("Lambda graph:", lambda_graph)

Lambda graph: 0.001


X: 62 x 2000 \
L_input: 62 x 62 \
lambda_graph: 0.001

#### Structure Funtion Loss Function

In [15]:
def laplacian_structure_loss(X_input, Z):
    """
    Compare the graph structure of the input data
    with the graph structure of its latent representation.
    """

    L_x = compute_laplacian(
        X_input.detach().cpu().numpy()
    )

    L_z = compute_laplacian(
        Z.detach().cpu().numpy()
    )

    L_x = torch.tensor(
        L_x,
        dtype=torch.float32,
        device=Z.device
    )

    L_z = torch.tensor(
        L_z,
        dtype=torch.float32,
        device=Z.device
    )

    return torch.mean((L_x - L_z) ** 2)

#### Train the Updated Model

In [16]:
import importlib
import src.models.autoencoder as autoencoder_module

importlib.reload(autoencoder_module)

AutoEncoder = autoencoder_module.AutoEncoder

print("AutoEncoder loaded successfully")

AutoEncoder loaded successfully


In [17]:
input_dim = X.shape[1]
hidden_dim = 128

lambda_l21 = 0.001
lambda_graph = 0.001

model = autoencoder_module.AutoEncoder(
    input_dim=input_dim,
    hidden_dim=hidden_dim,
    lambda_l21=lambda_l21
)

criterion = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

print(model)
print("Lambda l2,1:", lambda_l21)
print("Lambda graph:", lambda_graph)

AutoEncoder(
  (encoder_layer): Linear(in_features=2000, out_features=128, bias=True)
  (encoder): Sequential(
    (0): Linear(in_features=2000, out_features=128, bias=True)
    (1): ReLU()
  )
  (decoder): Sequential(
    (0): Linear(in_features=128, out_features=2000, bias=True)
  )
)
Lambda l2,1: 0.001
Lambda graph: 0.001


In [18]:
num_epochs = 50
l21_losses = []

model.train()

for epoch in range(num_epochs):
    epoch_loss = 0

    for (batch,) in train_loader:
        optimizer.zero_grad()

        output = model(batch)

        reconstruction_loss = criterion(output, batch)
        l21_loss = model.l21_penalty()

        # Get latent representation
        latent = model.encoder(batch)

        # Graph structure preservation loss
        graph_loss = laplacian_structure_loss(batch, latent)

        # Total loss
        loss = (
            reconstruction_loss
            + model.lambda_l21 * l21_loss
            + lambda_graph * graph_loss
        )

        loss.backward()
        optimizer.step()

        epoch_loss += reconstruction_loss.item()

    avg_loss = epoch_loss / len(train_loader)
    l21_losses.append(avg_loss)

    print(
        f"Epoch {epoch+1}/{num_epochs}, "
        f"Reconstruction Loss: {avg_loss:.4f}"
    )

Epoch 1/50, Reconstruction Loss: 2.4195
Epoch 2/50, Reconstruction Loss: 2.1525
Epoch 3/50, Reconstruction Loss: 1.8521
Epoch 4/50, Reconstruction Loss: 1.5845
Epoch 5/50, Reconstruction Loss: 1.3965
Epoch 6/50, Reconstruction Loss: 1.2992
Epoch 7/50, Reconstruction Loss: 1.2317
Epoch 8/50, Reconstruction Loss: 1.1734
Epoch 9/50, Reconstruction Loss: 1.1126
Epoch 10/50, Reconstruction Loss: 1.0525
Epoch 11/50, Reconstruction Loss: 0.9957
Epoch 12/50, Reconstruction Loss: 0.9550
Epoch 13/50, Reconstruction Loss: 0.8987
Epoch 14/50, Reconstruction Loss: 0.8555
Epoch 15/50, Reconstruction Loss: 0.8198
Epoch 16/50, Reconstruction Loss: 0.7881
Epoch 17/50, Reconstruction Loss: 0.7552
Epoch 18/50, Reconstruction Loss: 0.7163
Epoch 19/50, Reconstruction Loss: 0.6878
Epoch 20/50, Reconstruction Loss: 0.6630
Epoch 21/50, Reconstruction Loss: 0.6362
Epoch 22/50, Reconstruction Loss: 0.6221
Epoch 23/50, Reconstruction Loss: 0.6062
Epoch 24/50, Reconstruction Loss: 0.5926
Epoch 25/50, Reconstructi

#### Recalculate Feature Importance

In [19]:
weights_graph = model.encoder_layer.weight.detach()

feature_importance_graph = torch.norm(
    weights_graph,
    dim=0
)

print(
    "Number of feature scores:",
    feature_importance_graph.shape[0]
)

Number of feature scores: 2000


#### Feature Count sweep

In [20]:
y_np = df["label"].values

print(y_np.shape)

(62,)


In [23]:
from src.evaluation.evaluate import evaluate_clustering

print(evaluate_clustering)

<function evaluate_clustering at 0x78424a5f3380>


In [24]:
feature_counts = [50, 100, 150, 200, 250, 300]

graph_results = []

for k in feature_counts:

    top_indices = torch.topk(
        feature_importance_graph,
        k=k
    ).indices

    X_selected = X[:, top_indices]

    acc, nmi = evaluate_clustering(
        X_selected.cpu().numpy(),
        y_np
    )

    graph_results.append({
        "Dataset": "Colon",
        "Method": "L2,1 + Laplacian",
        "Feature_Count": k,
        "ACC": acc,
        "NMI": nmi
    })

    print(f"k={k}: ACC={acc:.4f}, NMI={nmi:.4f}")

k=50: ACC=0.5323, NMI=0.0006
k=100: ACC=0.5161, NMI=0.0000
k=150: ACC=0.5484, NMI=0.0062
k=200: ACC=0.5161, NMI=0.0000
k=250: ACC=0.5323, NMI=0.0017
k=300: ACC=0.5484, NMI=0.0062


In [25]:
feature_counts = [50, 100, 150, 200, 250, 300]

graph_results = []

for k in feature_counts:

    top_indices = torch.topk(
        feature_importance_graph,
        k=k
    ).indices

    X_selected = X[:, top_indices]

    acc, nmi = evaluate_clustering(
        X_selected.cpu().numpy(),
        y_np
    )

    graph_results.append({
        "Dataset": "Colon",
        "Method": "L2,1 + Laplacian",
        "Feature_Count": k,
        "ACC": acc,
        "NMI": nmi
    })

    print(f"k={k}: ACC={acc:.4f}, NMI={nmi:.4f}")

k=50: ACC=0.5323, NMI=0.0006
k=100: ACC=0.5161, NMI=0.0000
k=150: ACC=0.5484, NMI=0.0062
k=200: ACC=0.5161, NMI=0.0000
k=250: ACC=0.5323, NMI=0.0017
k=300: ACC=0.5484, NMI=0.0062


#### Automatically append Phase K results

In [26]:
# Convert Phase K results to DataFrame
graph_df = pd.DataFrame(graph_results)

# Add model version explicitly
graph_df["Model_Version"] = "L2,1 + Laplacian"

print(graph_df)

  Dataset            Method  Feature_Count       ACC       NMI  \
0   Colon  L2,1 + Laplacian             50  0.532258  0.000608   
1   Colon  L2,1 + Laplacian            100  0.516129  0.000014   
2   Colon  L2,1 + Laplacian            150  0.548387  0.006223   
3   Colon  L2,1 + Laplacian            200  0.516129  0.000014   
4   Colon  L2,1 + Laplacian            250  0.532258  0.001712   
5   Colon  L2,1 + Laplacian            300  0.548387  0.006223   

      Model_Version  
0  L2,1 + Laplacian  
1  L2,1 + Laplacian  
2  L2,1 + Laplacian  
3  L2,1 + Laplacian  
4  L2,1 + Laplacian  
5  L2,1 + Laplacian  


In [27]:
from pathlib import Path
results_dir = Path("reports/results")

In [28]:
# ---------- Master feature-selection results ----------

master_file = results_dir / "feature_selection_results.csv"

master_df = pd.read_csv(master_file)

# Add Model_Version to existing data if it doesn't exist
if "Model_Version" not in master_df.columns:
    master_df["Model_Version"] = master_df["Method"]

# Append Phase K
master_df = pd.concat(
    [master_df, graph_df],
    ignore_index=True
)

master_df.to_csv(master_file, index=False)

In [29]:
# ---------- Accuracy results ----------

accuracy_file = results_dir / "accuracy_results.csv"

accuracy_df = pd.read_csv(accuracy_file)

phase_k_acc = graph_df[
    ["Dataset", "Feature_Count", "ACC"]
].copy()

phase_k_acc["Method"] = "L2,1 + Laplacian"

phase_k_acc = phase_k_acc.rename(
    columns={"Feature_Count": "Features_Selected"}
)

# Put columns in the existing order
phase_k_acc = phase_k_acc[
    ["Dataset", "Method", "Features_Selected", "ACC"]
]

accuracy_df = pd.concat(
    [accuracy_df, phase_k_acc],
    ignore_index=True
)

accuracy_df.to_csv(accuracy_file, index=False)

In [30]:
# ---------- NMI results ----------

nmi_file = results_dir / "nmi_results.csv"

nmi_df = pd.read_csv(nmi_file)

phase_k_nmi = graph_df[
    ["Dataset", "Feature_Count", "NMI"]
].copy()

phase_k_nmi["Method"] = "L2,1 + Laplacian"

phase_k_nmi = phase_k_nmi.rename(
    columns={"Feature_Count": "Features_Selected"}
)

phase_k_nmi = phase_k_nmi[
    ["Dataset", "Method", "Features_Selected", "NMI"]
]

nmi_df = pd.concat(
    [nmi_df, phase_k_nmi],
    ignore_index=True
)

nmi_df.to_csv(nmi_file, index=False)

In [31]:
print(pd.read_csv("reports/results/feature_selection_results.csv").tail(6))
print(pd.read_csv("reports/results/accuracy_results.csv").tail(6))
print(pd.read_csv("reports/results/nmi_results.csv").tail(6))

   Dataset            Method  Feature_Count       ACC       NMI  \
10   Colon  L2,1 + Laplacian             50  0.532258  0.000608   
11   Colon  L2,1 + Laplacian            100  0.516129  0.000014   
12   Colon  L2,1 + Laplacian            150  0.548387  0.006223   
13   Colon  L2,1 + Laplacian            200  0.516129  0.000014   
14   Colon  L2,1 + Laplacian            250  0.532258  0.001712   
15   Colon  L2,1 + Laplacian            300  0.548387  0.006223   

       Model_Version  
10  L2,1 + Laplacian  
11  L2,1 + Laplacian  
12  L2,1 + Laplacian  
13  L2,1 + Laplacian  
14  L2,1 + Laplacian  
15  L2,1 + Laplacian  
   Dataset            Method  Features_Selected       ACC
11   Colon  L2,1 + Laplacian                 50  0.532258
12   Colon  L2,1 + Laplacian                100  0.516129
13   Colon  L2,1 + Laplacian                150  0.548387
14   Colon  L2,1 + Laplacian                200  0.516129
15   Colon  L2,1 + Laplacian                250  0.532258
16   Colon  L2,1 + La

In [32]:
import pandas as pd

print(pd.read_csv("reports/results/feature_selection_results.csv"))
print(pd.read_csv("reports/results/accuracy_results.csv"))
print(pd.read_csv("reports/results/nmi_results.csv"))

   Dataset            Method  Feature_Count       ACC       NMI  \
0    Colon  L2,1 Autoencoder             50  0.532258  0.000608   
1    Colon  L2,1 Autoencoder            100  0.532258  0.001712   
2    Colon  L2,1 Autoencoder            150  0.548387  0.003877   
3    Colon  L2,1 Autoencoder            200  0.548387  0.003877   
4    Colon  L2,1 Autoencoder            250  0.548387  0.003877   
5    Colon  L2,1 Autoencoder            300  0.548387  0.003877   
6    Colon  L2,1 Autoencoder            350  0.532258  0.000608   
7    Colon  L2,1 Autoencoder            400  0.532258  0.000608   
8    Colon  L2,1 Autoencoder            450  0.532258  0.000608   
9    Colon  L2,1 Autoencoder            500  0.532258  0.000608   
10   Colon  L2,1 + Laplacian             50  0.532258  0.000608   
11   Colon  L2,1 + Laplacian            100  0.516129  0.000014   
12   Colon  L2,1 + Laplacian            150  0.548387  0.006223   
13   Colon  L2,1 + Laplacian            200  0.516129  0.00001

## Feature Selection via l2,1-Norm — Comparison

For the Colon dataset, the L2,1 + Laplacian model was compared with the L2,1 Autoencoder baseline across 50, 100, 150, 200, 250, and 300 selected features.

The Laplacian-regularized model did not consistently improve clustering performance over the L2,1-only model. The best Phase K result was ACC = 0.5484 and NMI = 0.0062 at 150 and 300 features. The best Feature Selection via l2,1-Norm ACC was 0.5484, while its best NMI was 0.0039. Therefore, the Laplacian term did not provide a clear ACC improvement but produced a higher NMI for the best Structure Preservation: Input-Data Laplacian settings.

In [33]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   reports/results/accuracy_results.csv
	modified:   reports/results/feature_selection_results.csv
	modified:   reports/results/nmi_results.csv

no changes added to commit (use "git add" and/or "git commit -a")


In [34]:
!git add reports/results/accuracy_results.csv reports/results/feature_selection_results.csv reports/results/nmi_results.csv

In [37]:
!git commit -m "Add Laplacian structure preservation experiments"

[main 265b6a1] Add Laplacian structure preservation experiments
 3 files changed, 41 insertions(+), 23 deletions(-)
 rewrite reports/results/nmi_results.csv (89%)


In [38]:
!git push

Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 2 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 1.03 KiB | 1.03 MiB/s, done.
Total 7 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 2 local objects.
To https://github.com/cmshebeeb/SMFS-AutoEncoder.git
   1dd3af8..265b6a1  main -> main
